### PR TITLE
upate filter config provider to return reference wrapper of filter factory cb

### PR DESCRIPTION
**Pull request by wbpcode**
Signed-off-by: wbpcode <wangbaiping@corp.netease.com>

Commit Message: upate filter config provider to return reference wrapper of filter factory cb
Additional Description:

In the previous ecds implementation, we will always create a temp copy of `FactoryCb` which results in uncessary memory allocations and atomic operations (copy and destroy shared pointer in the `FactoryCb`).

Risk Level: Low.
Testing: N/A.
Docs Changes: N/A.
Release Notes: N/A.
Platform Specific Features: N/A.

### CODE REVIEW

**Comment by wbpcode** [comment]
If this is for some special design reason, feel free to close this PR.

**Comment by kyessenov** [comment]
LGTM. This was an oversight (some factories are expensive to copy).

**Comment by adisuissa** [comment]
cc @yanjunxiang-google

File: envoy/config/extension_config_provider.h [hunk 1]
```diff
@@ -32,7 +32,7 @@ template <class FactoryCallback> class ExtensionConfigProvider {
    * callback is the latest version of the extension configuration, and should
    * generally apply only to new requests and connections.
    */
-  virtual absl::optional<FactoryCallback> config() PURE;
+  virtual absl::optional<std::reference_wrapper<FactoryCallback>> config() PURE;
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply envoy/config/extension_config_provider.h:35]: get it.

File: source/common/filter/config_discovery_impl.h [hunk 2]
```diff
@@ -94,7 +94,12 @@ class DynamicFilterConfigProviderImpl : public DynamicFilterConfigProviderImplBa
 
   // Config::ExtensionConfigProvider
   const std::string& name() override { return DynamicFilterConfigProviderImplBase::name(); }
-  absl::optional<FactoryCb> config() override { return tls_->config_; }
+  absl::optional<std::reference_wrapper<FactoryCb>> config() override {
+    if (auto& optional_config = tls_->config_; optional_config.has_value()) {
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply source/common/filter/config_discovery_impl.h:98]: May we cannot. we can only get `const FacoryCb` from const optional_config.

**Comment by adisuissa** [comment]
/assign @htuch 
Please take a final pass

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**